File: 1019. Next Greater Node In Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def nextLargerNodes(self, head: ListNode) -> List[int]:
        result = []
        stack = []
        idx = 0
        while head:
            while stack and stack[-1][0] < head.val:
                val, i = stack.pop()
                result[i] = head.val
            stack.append((head.val, idx))
            result.append(0)
            idx += 1
            head = head.next
        return result

File: 1019. Next Greater Node In Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public int[] nextLargerNodes(ListNode head) {
        // push index into stack and value in list
        // update where we meet the val is larger than in stack
        Stack<Integer> st = new Stack<Integer>();
        List<Integer> ls = new ArrayList<Integer>();
        int index = 0;
        while (head != null) {
            ls.add(head.val);
            while (!st.empty() && ls.get(st.peek()) < head.val) {
                ls.set(st.pop(), head.val);
            }
            st.push(index);
            head = head.next;
            index++;
        }
        while (!st.empty()) {
            ls.set(st.pop(), 0);
        }
        int[] res = new int[ls.size()];
        for (int i=0; i<ls.size(); i++) {
            res[i] = ls.get(i);
        }
        return res;
    }
}